src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -1371,6 +1373,177 @@ int PeerReplayer::do_synchronize(const std::string &dir_root, const Snapshot &cu
+    return r;
+  }
+
+  BOOST_SCOPE_EXIT_ALL( (this)(&fh) ) {
+    post_sync_close_handles(fh);
+  };
+
+  // record that we are going to "dirty" the data under this directory root
+  auto snap_id_str{stringify(current.second)};
+  r = ceph_setxattr(m_remote_mount, dir_root.c_str(), "ceph.mirror.dirty_snap_id",
+                    snap_id_str.c_str(), snap_id_str.size(), 0);
+  if (r < 0) {
+    derr << ": error setting \"ceph.mirror.dirty_snap_id\" on dir_root=" << dir_root
+         << ": " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  struct ceph_statx cstx;
+  r = ceph_fstatx(m_local_mount, fh.c_fd, &cstx,
+                  CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                  CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                  AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
+  if (r < 0) {
+    derr << ": failed to stat snap=" << current.first << ": " << cpp_strerror(r)
+         << dendl;
+    return r;
+  }
+
+  ceph_snapdiff_info sd_info;
+  ceph_snapdiff_entry_t sd_entry;
+
+  //The queue of SyncEntry items (directories) to be synchronized.
+  //We follow a breadth first approach here based on the snapdiff output.
+  std::queue<SyncEntry> sync_queue;
+
+  //start with initial/default entry
+  std::string epath = ".", npath = "", nabs_path = "", nname = "";
+  sync_queue.emplace(SyncEntry(epath, cstx));
+
+  while (!sync_queue.empty()) {
+    if (should_backoff(dir_root, &r)) {
+      dout(0) << ": backing off r=" << r << dendl;
+      break;
+    }
+    r = pre_sync_check_and_open_handles(dir_root, current, prev, &fh);
+    if (r < 0) {
+      dout(5) << ": cannot proceed with sync: " << cpp_strerror(r) << dendl;
+      return r;
+    }
+
+    dout(20) << ": " << sync_queue.size() << " entries in queue" << dendl;
+    const auto &queue_entry = sync_queue.front();
+    epath = queue_entry.epath;
+    dout(20) << ": syncing entry, path=" << epath << dendl;
+    r = ceph_open_snapdiff(fh.p_mnt, dir_root.c_str(), epath.c_str(),
+                           stringify((*prev).first).c_str(), current.first.c_str(), &sd_info);
+    if (r != 0) {
+      derr << ": failed to open snapdiff, r=" << r << dendl;
+      return r;
+    }
+    while (0 < (r = ceph_readdir_snapdiff(&sd_info, &sd_entry))) {
+      if (r < 0) {
+        derr << ": failed to read directory=" << epath << dendl;
+        ceph_close_snapdiff(&sd_info);
+        return r;
+      }
+
+      //New entry found
+      nname = sd_entry.dir_entry.d_name;
+      if ("." == nname || ".." == nname)
+        continue;
+      // create path for the newly found entry
+      npath = entry_diff_path(epath, nname);
+      nabs_path = entry_diff_path(dir_root, npath);
+
+      r = ceph_statx(sd_info.cmount, nabs_path.c_str(), &cstx,
+                     CEPH_STATX_MODE | CEPH_STATX_UID | CEPH_STATX_GID |
+                     CEPH_STATX_SIZE | CEPH_STATX_ATIME | CEPH_STATX_MTIME,
+                     AT_STATX_DONT_SYNC | AT_SYMLINK_NOFOLLOW);
```

Comment:
Shouldn't this absolute path point to the path in the current snapshot directory? Is it doing that? For example If we are mirroring `/root` directory and directory is `/root/x` then shouldn't be the path here `/root/.snap/<cur_snap>/x`?

Reply:
It's the same thing.
`/root/` and `/root/.snap/<cur_snap>/` are the same.

Checkout the previous [do_synchronize](https://github.com/joscollin/ceph/blob/main/src/tools/cephfs_mirror/PeerReplayer.cc#L1234) code.

Reply:
Why it should be same, let's suppose take this scenario,
1. do some changes take snap_1
2. do some changes take snap_2
3. do some changes and add the directory for mirroring

In that case `/root/.snap/snap_2` and `/root/`, does it represent same picture? In previous case I think relative path is used. Let me know if I am wrong

Reply:
The sync happens only when we snapshot local directory, not when you add the directory for mirroring.

Reply:
Sorry, I think you misunderstood me, I am saying initially root directory is not added for mirroring,
1. Did some changes took snap_1
2. Did some changes took snap_2
3. Did some changes and add the `/root` directory for mirroring

In that case mirror will start when I am adding `/root` is added for mirroring,
So mirror will happen like this, first snap_1 will be propagated as there is no previous snapshot. Then delta between snap_2 and snap_1 will be propagated but problem is you did some changes in the `/root`. So assumption that `/root/.snap/snap_2` and `/root` is not correct isn't it?

Reply:
That's a potential bug then, as we could create a dir/file, snapshot it and then delete the dir/file and stat it. Thus the resulting local & remote snapshots would be different. @sajibreadd Please create a tracker for this and assign to me.

The snapdiff api just shows the diff and doesn't show the operations done for an entry.  So we need to identify it by comparing. IMO, the api should've been improved.

I don't know how this issue was handled before snapdiff was merged. @vshankar Any thoughts on this?

Reply:
I think previously It was not causing any problem as relative path was taken, so relative path was `./x`, fh->p_fd was open on the `prev_snapshot_path`. So path resolved to the valid expected path.

Reply:
The path used in libcephfs APIs should be relative to the mount, i.e. something like `./path/to/dir`.